scripts/opentitan/gdbreplay.py

```diff
@@ -0,0 +1,899 @@
+
+           :return: the address width in bits (not bytes!)
+        """
+        return self._elf.elfclass if self._elf else 0
+
+    @property
+    def entry_point(self) -> Optional[int]:
+        """Provide the entry point of the application, if any.
+
+           :return: the entry point address
+        """
+        return self._elf and self._elf.header.get('e_entry', None)
+
+    @property
+    def raw_size(self) -> int:
+        """Provide the size of the Secure Boot Header section, if any.
+
+           :return: the data/payload size in bytes
+        """
+        if not self._payload_size:
+            self._payload_address, self._payload_size = self._parse_segments()
+        return self._payload_size
+
+    @property
+    def load_address(self) -> int:
+        """Provide the first destination address on target to copy the
+           application blob.
+
+           :return: the load address
+        """
+        if not self._payload_address:
+            self._payload_address, self._payload_size = self._parse_segments()
+        return self._payload_address
+
+    @property
+    def blob(self) -> bytes:
+        """Provide the application blob, i.e. the whole loadable binary.
+
+           :return: the raw application binary.
+        """
+        if not self._payload:
+            self._payload = self._build_payload()
+        if len(self._payload) != self.raw_size:
+            raise RuntimeError('Internal error: size mismatch')
+        return self._payload
+
+    def _loadable_segments(self) -> Iterator[Segment]:
+        """Provide an iterator on segments that should be loaded into the final
+           binary.
+        """
+        if not self._elf:
+            raise RuntimeError('No ELF file loaded')
+        for segment in sorted(self._elf.iter_segments(),
+                              key=lambda seg: seg['p_paddr']):
+            if segment['p_type'] not in ('PT_LOAD', ):
+                continue
+            if not segment['p_filesz']:
+                continue
+            yield segment
+
+    def _parse_segments(self) -> Tuple[int, int]:
+        """Parse ELF segments and extract physical location and size.
+
+           :return: the location of the first byte and the overall payload size
+                    in bytes
+        """
+        size = 0
+        phy_start = None
+        for segment in self._loadable_segments():
+            seg_size = segment['p_filesz']
+            if not seg_size:
+                continue
+            phy_addr = segment['p_paddr']
+            if phy_start is None:
+                phy_start = phy_addr
+            else:
+                if phy_addr > phy_start+size:
+                    self._log.debug('fill gap with previous segment')
+                    size = phy_addr-phy_start
+            size += seg_size
+        if phy_start is None:
+            raise ValueError('No loadable segment found')
+        return phy_start, size
+
+    def _build_payload(self) -> bytes:
```

Comment:
`_parse_segments` and `_build_payload` look very similar, maybe factorize common code?

Reply:
We do not want to build the full object in RAM when we only need to compute the final size.